### PR TITLE
Menu auto hide

### PR DESCRIPTION
**Pull request by jwrdegoede**
Hi Peter, Javier,

Here is a series of grub patches for the boot experience improvements I've been working on.

The first patch is one you already know, which stops grub from needlessly switching the EFI display to text mode when there is no output. This version fixes a few small compiler warnings (I removed a bunch of __unused attributes which should stay) and is otherwise unmodified

Patches 2-4 add getkeystatus() support to the EFI console driver and furhter preparation work for patch 5, which adds F8 and keeping shift pressed as extra methods to unhide the menu

Patch 6 modifies the grub-editenv util so that it can increment integer value grubenv variables, we need this for the boot_indeterminate count, the plan is e.g. to add a "grub2-editenv - incr boot_indeterminate" call to /usr/libexec/selinux/selinux-autorelabel

Patch 7 adds a new file dropping into grub.d adding the menu auto-hide stuff to grub.cfg. I put this in a separate file for easier rebasing on top of newer upstream grub versions

Patch 8 is taken 1:1 from Ubuntu, it adds a "System setup" menu entry to enter the EFI firmware setup, which is useful when using fastboot. Note patch 7 also allows doing this by setting "fwsetup_once=1" in the grunenv, but I believe it is good to have multiple ways to do this for systems where the firmware itself does not check for keypresses.

Please review and cherry-pick the ones which you think are ready for merging. Then I will rebase and address comments on the other patches.

Regards,

Hans

### CODE REVIEW

File: grub-core/kern/term.c [hunk 1]
```diff
@@ -138,6 +138,22 @@ grub_getkeystatus (void)
   return status;
 }
 
+int
+grub_key_is_interrupt (int key)
+{
+  /* ESC sometimes is the BIOS setup hotkey and may be hard to discover, also
+     check F8, which was the key to get the Windows bootmenu for a long time. */
+  if (key == GRUB_TERM_ESC || key == GRUB_TERM_KEY_F8)
```
Comment (martinezjavier) [inline grub-core/kern/term.c]: Something that I missed yesterday, in the upstream discussion for this patch someone mentioned that F4 may be a better option than F8 and the the maintainer seems to agrees. So I guess that in v2 you will change that and also add the SHIFT pressed option that wasn't in your original patch?

So I think that we should also use F4 downstream, to be consistent with what would be upstream grub2 behaviour after your patch and also so we can drop this patch from the package once yours land and we do a rebase.

**Comment by jwrdegoede** [comment]
Hi,

On 08-06-18 10:30, Javier Martinez Canillas wrote:
> *@martinezjavier* commented on this pull request.
> 
> ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
> 
> In grub-core/kern/term.c <https://github.com/rhboot/grub2/pull/11#discussion_r193987038>:
> 
>> @@ -138,6 +138,22 @@ grub_getkeystatus (void)
>     return status;
>   }
>   
> +int
> +grub_key_is_interrupt (int key)
> +{
> +  /* ESC sometimes is the BIOS setup hotkey and may be hard to discover, also
> +     check F8, which was the key to get the Windows bootmenu for a long time. */
> +  if (key == GRUB_TERM_ESC || key == GRUB_TERM_KEY_F8)
> 
> Something that I missed yesterday, in the upstream discussion for this patch someone mentioned that F4 may be a better option than F8 and the the maintainer seems to agrees. So I guess that in v2 you will change that and also add the SHIFT pressed option that wasn't in your original patch?
> 
> So I think that we should also use F4 downstream, to be consistent with what would be upstream grub2 behaviour after your patch and also so we can drop this patch from the package once yours land and we do a rebase.

You are right that the upstream discussion suggested F4. But I discussed this
with Alberto and the problem is discoverability, F4 is very hard to figure
out for users. Where as F8 is something which they may know from Windows
(if they ever used XP). So as a result from the discussion with Alberto
I've decided to stick with F8.

I might take another shot at getting this upstream with F8, but my
previous attempt at upstreaming things like my:

"Optionally print less messages at boot"

Where I got unrealistic feedback such as suggestions that rather
then compile time this should be runtime with stuffing some
flag in the boot sector for classic BIOS boot has not exactly
made me enthusiastic about upstreaming grub patches.

I would love for our delta with upstream to be smaller, but even after
wrapping the silent stuff in a compile time option it still gets a
lot of pushback from upstream. Despite this being something which
a whole bunch of distros have been patching downstream for
10 years now. *10* years !!

Likewise the hold SHIFT down behavior has been submitted upstream
in 2009 already:

https://lists.gnu.org/archive/html/grub-devel/2009-08/msg00718.html

TL;DR: I would love to make our delta with upstream smaller, but
I really don't want to spend way too much time bikeshedding with
upstream about this.

Regards,

Hans

**Comment by martinezjavier** [comment]
> You are right that the upstream discussion suggested F4. But I discussed this
> with Alberto and the problem is discoverability, F4 is very hard to figure
> out for users. Where as F8 is something which they may know from Windows
> (if they ever used XP). So as a result from the discussion with Alberto
> I've decided to stick with F8.

Ah Ok, I didn't know this was already discussed. Let's stick with F8 then.

> Where I got unrealistic feedback such as suggestions that rather
> then compile time this should be runtime with stuffing some
> flag in the boot sector for classic BIOS boot has not exactly
> made me enthusiastic about upstreaming grub patches.

Yeah, I saw that too. Probably now you understand why we carry so many patches on top of mainline grub2...

**Comment by jwrdegoede** [comment]
Hi,

So I've just pushed an updated version of the menu-auto-hide branch this pullreq is for. I've changed 2 things:

1) Changed the "Add auto-hide menu support" to no longer check for shutdown_success and dropped fwsetup_once support from it since systemctl reboot --firmware can already do that.

2) Added a new "Add grub-set-bootflag utility" commit, this is a little helper intended to be pkexec-ed so it only supports a hardcoded lists of flags to set in the env (boot_success, menu_show_once) . It also contains sample systemd user session timer and service files as well as a policykit policy file, see the commit message for how to install these, with these in place boot_success will automatically be set 2 minutes after the user has logged in.

I believe that this is ready to be merged now (after review of the new patch). It would be good to move forward with this.

Regards,

Hans

**Comment by jwrdegoede** [comment]
Note I'm preparing a test rpm build of grub2 from my menu-auto-hide branch, which includes spec file changes to put the systemd user session and polkit policy files in the right place. So if you want to give this a test run, it is probably best if you wait a bit then you can test this by simply installing a rpm.

File: util/grub-set-bootflag.c [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    {
+      fprintf (stderr, "Invalid bootflag: '%s'\n", argv[1]);
+      usage();
+      return 1;
+    }
+
+  bootflag = bootflags[i];
+  len = strlen (bootflag);
+
+  f = fopen (GRUBENV, "r");
```
Comment (martinezjavier) [inline util/grub-set-bootflag.c]: The tools that are using `grub_envblk_open()` will check if the file starts with `GRUB_ENVBLK_SIGNATURE` to take the environment block as valid. I know that you are not using any grub related code to ease auditing, but I wonder if you shouldn't check for that as well to be consistent with the other tools.

**Comment by jwrdegoede** [comment]
Hi Javier,

I agree with both your comments in your review of the grub-set-bootlfag utility, I've just pushed a new version fixing this.

If you want to test the polkit / systemd integration I've rpms of this new version here:
https://fedorapeople.org/~jwrdegoede/grub2-2.02-37.hdg1.fc29.x86_64/

If you do:

systemctl --user list-timers

After logging into a GNOME3 session, then you should see a grub-boot-success.timer there, once that has expired you "sudo grub2-editenv list" should show boot_success=1

Regards,

Hans

**Comment by martinezjavier** [comment]
> If you want to test the polkit / systemd integration I've rpms of this new version here:
> https://fedorapeople.org/~jwrdegoede/grub2-2.02-37.hdg1.fc29.x86_64/

I've also tested your packages and worked as expected.

```
$ systemctl --user list-timers
NEXT                         LEFT    LAST PASSED UNIT                    ACTIVATES
Wed 2018-06-13 17:06:56 BST  2s left n/a  n/a    grub-boot-success.timer grub-boot-success.service

1 timers listed.
Pass --all to see loaded but inactive timers, too.

$ grub2-editenv list                                                                                                                                                                                           
saved_entry=Fedora (4.16.3-301.fc28.x86_64) 28 (Workstation Edition)

$ systemctl --user list-timers
NEXT LEFT LAST                         PASSED  UNIT                    ACTIVATES
n/a  n/a  Wed 2018-06-13 17:06:57 BST  9ms ago grub-boot-success.timer grub-boot-success.service

1 timers listed.
Pass --all to see loaded but inactive timers, too.

$ grub2-editenv list                                                                                                                                                                                           
saved_entry=Fedora (4.16.3-301.fc28.x86_64) 28 (Workstation Edition)
boot_success=1
```
